### PR TITLE
FM2-305: Create Integration Tests for Practitioner.

### PR DESCRIPTION
**Pull request by Akayeshmantha**
## Description of what I changed
Create Integration Tests for Practitioner

## Issue I worked on
see https://issues.openmrs.org/browse/FM2-305

## Checklist: I completed these to help reviewers :)
- [x ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x ] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [ x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x ] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x ] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,480 @@
...
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.SEARCHSET));
+		assertThat(results.hasEntry(), is(true));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasResource(hasProperty("nameFirstRep", hasProperty("family", startsWith("Doe"))))));
+		assertThat(entries, containsInRelativeOrder(
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java:433]: Can we use some other search parameter such that we get multiple entries in the search results? With this, we are not able to justify if sorting works correctly as search returns just 1 result.

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java:476]: Ditto

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+	public void shouldReturnNotFoundWhenUpdatingNonExistentPractitionerAsXML() throws Exception {
+		// get the existing record
+		MockHttpServletResponse response = get("/Practitioner/" + PRACTITIONER_UUID).accept(FhirMediaTypes.JSON).go();
+		Practitioner practitioner = readResponse(response);
+		
+		// update the existing record
+		practitioner.setId(WRONG_PRACTITIONER_UUID);
+		
+		// send the update to the server
+		response = put("/Practitioner/" + WRONG_PRACTITIONER_UUID).jsonContent(toJson(practitioner)).go();
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java:357]:
```suggestion
		response = put("/Practitioner/" + WRONG_PRACTITIONER_UUID).xmlContext(toXML(practitioner)).go();
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+		
+		assertThat(response, isNotFound());
+		assertThat(response.getContentType(), is(FhirMediaTypes.JSON.toString()));
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java:360]:
```suggestion
		assertThat(response.getContentType(), is(FhirMediaTypes.XML.toString()));
```

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,480 @@
...
+		Bundle results = readBundleResponse(response);
+		
+		assertThat(results, notNullValue());
+		assertThat(results.getType(), equalTo(Bundle.BundleType.SEARCHSET));
+		assertThat(results.hasEntry(), is(true));
+		
+		List<Bundle.BundleEntryComponent> entries = results.getEntry();
+		
+		assertThat(entries, everyItem(hasResource(hasProperty("nameFirstRep", hasProperty("family", startsWith("Doe"))))));
+		assertThat(entries, containsInRelativeOrder(
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java:476]: Ditto

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,480 @@
...
```
Comment (varung-31) [inline integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r4/PractitionerFhirResourceProviderIntegrationTest.java:433]: Ditto

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/PractitionerTranslatorProviderImpl.java [hunk 7]
```diff
@@ -72,9 +75,25 @@ public Provider toOpenmrsType(Provider existingProvider, Practitioner practition
 		
 		existingProvider.setUuid(practitioner.getId());
 		
-		Validate.notEmpty(practitioner.getIdentifier(), "Practitioner Identifier cannot be empty");
+		notEmpty(practitioner.getIdentifier(), "Practitioner Identifier cannot be empty");
 		existingProvider.setIdentifier(practitioner.getIdentifier().get(0).getValue());
 		
+		if (existingProvider.getPerson() == null) {
```
Comment (varung-31) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/PractitionerTranslatorProviderImpl.java:81]: Could you please add some tests for the translator changes?

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,480 @@
...
+	public void shouldReturnNotFoundWhenUpdatingNonExistentPractitionerAsXML() throws Exception {
+		// get the existing record
+		MockHttpServletResponse response = get("/Practitioner/" + PRACTITIONER_UUID).accept(FhirMediaTypes.JSON).go();
+		Practitioner practitioner = readResponse(response);
+		
+		// update the existing record
+		practitioner.setId(WRONG_PRACTITIONER_UUID);
+		
+		// send the update to the server
+		response = put("/Practitioner/" + WRONG_PRACTITIONER_UUID).jsonContent(toJson(practitioner)).go();
```
[the remark this answers is not in the document]
Comment (Akayeshmantha) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java:357]: oh bad mistakes :+1:

File: integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java (hunk 1 above)
Comment (Akayeshmantha) [reply integration-tests/src/test/java/org/openmrs/module/fhir2/providers/r3/PractitionerFhirResourceProviderIntegrationTest.java]: Hi @varung-31  sort was not implemented, in-service class. Removing sort tests.

**Review by ibacher — APPROVED** [review]

**Pull request merged**